File: fsharp/isbn-verifier/IsbnVerifier.fs
module IsbnVerifier

open System.Text.RegularExpressions

let private normalize (isbn: string) = isbn.Replace("-", "")

let private hasValidCharacters isbn = Regex.IsMatch(isbn, @"^\d{9}[\dX]$")

let private digitToValue index digit =
    let indexFactor = index + 1

    match digit with
    | 'X' -> 10 * indexFactor
    | _   -> (int digit - int '0') * indexFactor

let private checksum isbn =
    isbn
    |> Seq.rev
    |> Seq.mapi digitToValue
    |> Seq.sum

let isValid isbn = 
    let normalizedIsbn = normalize isbn
    hasValidCharacters normalizedIsbn && checksum normalizedIsbn % 11 = 0

File: python/isbn-verifier/isbn_verifier.py
from re import match


def is_valid(isbn):
    m = match(r'^\d{9}[\dX]$', isbn.replace('-', ''))

    if not m:
        return False

    return sum(10 if char == 'X' else int(char) * (10 - i) for i, char in enumerate(m.string)) % 11 == 0
